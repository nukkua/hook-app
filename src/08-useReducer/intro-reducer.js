const initialState = [
	{
		id: 1,
		description: 'Leer Clean Code',
		done: false
	}
];
const stateReducer = (state = initialState, action = {}) => {
	if (action.type === '[TODO] add todo') {
		return [...state, action.payload];
	}

	return state;
};

let todo = stateReducer();
console.log(todo);

const newTodo = {
	id: 2,
	description: 'Responder a camila',
	done: false
};

const addTodoAction = {
	type: '[TODO] add todo',
	payload: newTodo
};

todo = stateReducer(todo, addTodoAction);
console.log(todo);
