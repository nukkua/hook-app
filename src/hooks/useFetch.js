import { useEffect, useState } from 'react';

export const useFetch = (url) => {
	const [state, setState] = useState({
		data: null,
		isLoading: true,
		hasError: null
	});

	const getFetch = async () => {
		setState({
			...state,
			isLoading: true
		});
		try {
			const resp = await fetch(url);
			if (!resp.ok) throw new Error('Error fetching data');
			const data = await resp.json();

			setState({
				data: data,
				isLoading: false,
				hasError: null
			});
		} catch (e) {
			setState({
				data: null,
				isLoading: false,
				hasError: e || 'something went wrong!'
			});
		}
	};

	useEffect(() => {
		getFetch();
	}, [url]);

	return {
		data: state.data,
		isLoading: state.isLoading,
		hasError: state.hasError
	};
};
