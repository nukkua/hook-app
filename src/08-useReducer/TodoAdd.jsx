import { useForm } from '../hooks/useForm';

export const TodoAdd = ({ onAddNewTodo }) => {
	const { description, handleInputChange, onResetForm } = useForm({
		description: ''
	});
	const handleSubmit = (event) => {
		event.preventDefault();
		if (description.length <= 1) return;
		const newTodo = {
			id: new Date().getTime(),
			done: false,
			description
		};
		onAddNewTodo(newTodo);
		onResetForm();
	};
	return (
		<form action="" onSubmit={handleSubmit}>
			<input
				value={description}
				onChange={handleInputChange}
				type="text"
				placeholder="What to do?"
				className="form-control"
				name="description"
			/>
			<button type="submit" className="btn btn-outline-primary mt-1">
				Agregar
			</button>
		</form>
	);
};
