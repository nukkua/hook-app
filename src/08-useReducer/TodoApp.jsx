import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';
import { useTodo } from '../hooks/useTodo';

export const TodoApp = () => {
	const { todos, handleNewTodo, handleDeleteTodo, handleToggleTodo, todosCount, pedingTodosCount } =
		useTodo();
	return (
		<>
			<h1>
				Todo App: {todosCount} <small>pendientes: {pedingTodosCount}</small>
			</h1>
			<hr />
			<div className="row">
				<div className="col-7">
					<TodoList todos={todos} onDeleteTodo={handleDeleteTodo} onToggleTodo={handleToggleTodo} />
				</div>
				<div className="col-5">
					<h4>Add TODO</h4>
					<hr />
					<TodoAdd onAddNewTodo={handleNewTodo} />
				</div>
			</div>
		</>
	);
};
