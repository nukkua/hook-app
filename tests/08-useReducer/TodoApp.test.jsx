import { render, screen } from '@testing-library/react';
import { useTodo } from '../../src/hooks/useTodo';
import { TodoApp } from '../../src/08-useReducer/TodoApp';

jest.mock('../../src/hooks/useTodo');

describe('Pruebas en el componente <TodoApp>', () => {
	useTodo.mockReturnValue({
		todos: [
			{ id: 1, description: 'Hacer Popo', done: false },
			{ id: 2, description: 'Hacer Pis', done: true }
		],
		todosCount: 2,
		pedingTodosCount: 1,
		handleDeleteTodo: jest.fn(),
		handleToggleTodo: jest.fn(),
		handleNewTodo: jest.fn()
	});

	test('debe de hacer match con el snapshot', () => {
		const { container } = render(<TodoApp />);
		expect(container).toMatchSnapshot();
	});

	test('debe de mostrar el componente correctamente', () => {
		render(<TodoApp />);
		expect(screen.getByText('Hacer Popo')).toBeTruthy();
		expect(screen.getByText('Hacer Pis')).toBeTruthy();
		expect(screen.getByRole('textbox')).toBeTruthy();
	});
});
