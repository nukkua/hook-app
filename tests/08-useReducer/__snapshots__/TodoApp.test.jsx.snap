// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <TodoApp> debe de hacer match con el snapshot 1`] = `
<div>
  <h1>
    Todo App: 
    2
     
    <small>
      pendientes: 
      1
    </small>
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-7"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            aria-label="span"
            class="align-self-center "
          >
            Hacer Popo
          </span>
          <button
            class="btn btn-danger"
          >
            Borrar
          </button>
        </li>
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            aria-label="span"
            class="align-self-center text-decoration-line-through"
          >
            Hacer Pis
          </span>
          <button
            class="btn btn-danger"
          >
            Borrar
          </button>
        </li>
      </ul>
    </div>
    <div
      class="col-5"
    >
      <h4>
        Add TODO
      </h4>
      <hr />
      <form
        action=""
      >
        <input
          class="form-control"
          name="description"
          placeholder="What to do?"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
          type="submit"
        >
          Agregar
        </button>
      </form>
    </div>
  </div>
</div>
`;
